### PR TITLE
fix: copy handler migrated to latest api

### PR DESCRIPTION
**Pull request by alihassan143**
#189 
#243 
[#2888](https://github.com/AppFlowy-IO/AppFlowy/issues/2888)

### CODE REVIEW

**Comment by alihassan143** [comment]
https://github.com/AppFlowy-IO/appflowy-editor/assets/71745182/d777f5f0-bd27-492f-b678-ed4291b4ae0d

@LucasXu0

**Comment by alihassan143** [comment]
https://github.com/AppFlowy-IO/appflowy-editor/assets/71745182/18b5897a-931d-48db-b138-b0f2f588a800

**Macos**

**Comment by alihassan143** [comment]
https://github.com/AppFlowy-IO/appflowy-editor/assets/71745182/b6ee84d8-8433-44c3-aaa0-2d5b6ce24a47

**Android real device**

**Comment by LucasXu0** [comment]
https://github.com/AppFlowy-IO/appflowy-editor/assets/11863087/b03a8f74-c7b7-48c0-8088-02a3b14d6e5e

@alihassan143 please check this video

**Comment by alihassan143** [comment]
@LucasXu0 paste handler is not migrated yeh so i need to migrate the paste handler also

**Comment by alihassan143** [comment]
@LucasXu0 copy command handler is  used in desktop for paste text

**Comment by alihassan143** [comment]
https://github.com/AppFlowy-IO/assets/71745182/021fbcca-5d72-4504-8b04-0068b44bc3ad

@LucasXu0

**Comment by alihassan143** [comment]
@LucasXu0  will you review this now its ready for review

**Comment by LucasXu0** [comment]
@alihassan143 Sure. I checked your video and found an issue.

If the text consists of just a single line, we should insert it within the existing paragraph.

For example, ABC|(<- paste a new line here)EFG, the result should be ABCXXXEFG instead of 
ABC
XXX
EFG

**Comment by alihassan143** [comment]
> @alihassan143 Sure. I checked your video and found an issue.
> 
> If the text consists of just a single line, we should insert it within the existing paragraph.
> 
> For example, ABC|(<- paste a new line here)EFG, the result should be ABCXXXEFG instead of ABC XXX EFG

its done

**Comment by LucasXu0** [comment]
FYI, @alihassan143. It takes time to review. Too many files were modified at once.

**Comment by alihassan143** [comment]
https://github.com/AppFlowy-IO@LucasXu0 appflowy-editor/assets/71745182/32948130-230c-49fa-a4db-5f880b2feae2

@LucasXu0 
kindly review this pr

**Comment by alihassan143** [comment]
@LucasXu0  pr is ready for review will you review it?

**Comment by alihassan143** [comment]
@Xazin  @LucasXu0  changes applied

**Comment by LucasXu0** [comment]
Great! I will do some refactor and optimize the logic.

**Comment by alihassan143** [comment]
@LucasXu0  this should be included in 1.2 version because lot of things changes but copy paste will not works on 1.2 version

**Comment by alihassan143** [comment]
@LucasXu0  you will rebase this or i will try?

File: test/new/command/copy_command_test.dart [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  await editor.pressKey(
+    key: LogicalKeyboardKey.keyC,
+    isControlPressed: Platform.isWindows || Platform.isLinux,
+    isMetaPressed: Platform.isMacOS,
+  );
+  await editor.pressKey(
+    key: LogicalKeyboardKey.backspace,
+    isControlPressed: Platform.isWindows || Platform.isLinux,
+    isMetaPressed: Platform.isMacOS,
+  );
```
Comment (MayurSMahajan) [inline test/new/command/copy_command_test.dart:40]: @alihassan143 Please check here that when you press backspace, the document is actually cleared.
Meaning that check if document is empty when you select all, and press backspace.

File: test/new/command/copy_command_test.dart [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+void main() async {
+  setUpAll(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+  });
+
+  group('copy_command_test.dart', () {
+    testWidgets('Presses Command + A in small document and copy text',
+        (tester) async {
+      await _testhandleCopy(tester, Document.fromJson(paragraphdata));
```
Comment (MayurSMahajan) [inline test/new/command/copy_command_test.dart:17]: Please follow camelCase name convention and rename `_testhandleCopy` to `_testHandleCopy`

File: test/new/command/copy_command_test.dart [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    isMetaPressed: Platform.isMacOS,
+  );
+  expect(
+    editor.editorState.document.toJson(),
+    document.toJson(),
+  );
+  await editor.dispose();
+}
+
+const paragraphText =
```
Comment (MayurSMahajan) [inline test/new/command/copy_command_test.dart:53]: Remove `paragraphText` if you are not using it.

File: test/new/command/cut_command_test.dart [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+import '../../new/infra/testable_editor.dart';
+
+void main() async {
+  setUpAll(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+  });
+  group('cut_command_test.dart', () {
+    testWidgets('update selection and execute cut command', (tester) async {
+      await _testcutHandle(tester, Document.fromJson(paragraphdata));
```
Comment (MayurSMahajan) [inline test/new/command/cut_command_test.dart:15]: Please follow camelCase name convention and rename `_testcutHandle` to `_testCutHandle`

File: test/new/command/paste_command_test.dart [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+void main() async {
+  setUpAll(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+  });
+  group('paste_command_test.dart', () {
+    testWidgets(
+        'Presses Command + A in small document and copy text and paste text',
+        (tester) async {
+      await _testhandleCopyPaste(tester, Document.fromJson(paragraphdata));
```
Comment (MayurSMahajan) [inline test/new/command/paste_command_test.dart:17]: Camel Case naming convention

File: test/new/command/paste_command_test.dart [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  await editor.pressKey(
+    key: LogicalKeyboardKey.keyP,
+    isControlPressed: Platform.isWindows || Platform.isLinux,
+    isMetaPressed: Platform.isMacOS,
+  );
+  expect(
+    editor.editorState.document.toJson(),
+    document.toJson(),
+  );
+
```
Comment (MayurSMahajan) [inline test/new/command/paste_command_test.dart:53]: Can you write a test to check the document after calling Paste Command multiple times? In fact, you don't need to create new test file, just within this file only, you can press the Paste command again and check if the document gets changed or not.

**Review by MayurSMahajan — CHANGES_REQUESTED** [review]

File: lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/cut_command.dart [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:flutter/material.dart';
+
+/// Copy.
```
Comment (Xazin) [inline lib/src/editor/editor_component/service/shortcuts/command_shortcut_events/cut_command.dart:4]:
```suggestion
/// Cut
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 8]
```diff
@@ -1,4 +1,13 @@
 import 'package:appflowy_editor/appflowy_editor.dart';
+import 'package:flutter/widgets.dart';
+
+int _textLengthOfNode(Node node) {
+  if (node.delta != null) {
+    return node.delta!.length;
+  }
+
+  return 0;
+}
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:10]:
```suggestion
int _textLengthOfNode(Node node) => node.delta?.length ?? 0;
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 9]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
   }
 }
+
+void pasteHTML(EditorState editorState, String html) {
+  final selection = editorState.selection?.normalized;
+  if (selection == null) {
+    return;
+  }
+
+  assert(selection.isCollapsed);
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:92]: Shouldnt this be
```suggestion
  if (selection == null || selection.isCollapsed) {
    return;
  }
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 10]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+
+Selection _computeSelectionAfterPasteMultipleNodes(
+  EditorState editorState,
+  List<Node> nodes,
+) {
+  final currentSelection = editorState.selection!;
+  final currentCursor = currentSelection.start;
+  final currentPath = [...currentCursor.path];
+  currentPath[currentPath.length - 1] += nodes.length;
+  int lenOfLastNode = _textLengthOfNode(nodes.last);
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:116]: final

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 11]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+        type: 'page',
+        children: nodes,
+      ),
+    ),
+  );
+  AppFlowyClipboard.setData(
+    text: text,
+    html: html.isEmpty ? null : html,
+  );
+  return;
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:144]: Redundant return

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 12]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+}
+
+void _pasteMultipleLinesInText(
+  EditorState editorState,
+  int offset,
+  List<Node> nodes,
+) {
+  final tb = editorState.transaction;
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:152]: What is tb? (Hint: better naming)

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 13]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+
+  final selectionNode = editorState.getNodesInSelection(selection!);
+  if (selectionNode.length == 1) {
+    final node = selectionNode.first;
+    final delta = node.delta;
+    if (delta == null) {
+      tb.afterSelection = afterSelection;
+      tb.insertNodes(afterSelection.end.path, nodes);
+      editorState.apply(tb);
+    }
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:165]: Seems you don't need delta variable much.

```suggestion
    final node = selectionNode.first;
    if (node.delta == null) {
      tb.afterSelection = afterSelection;
      tb.insertNodes(afterSelection.end.path, nodes);
      editorState.apply(tb);
    }
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 14]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+    ], [
+      firstnode,
+      ...nodes,
+      if (afternode != null &&
+          afternode.delta != null &&
+          afternode.delta!.isNotEmpty)
+        afternode,
+    ]);
+    editorState.apply(tb);
+    return;
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:216]: Redundant return

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 15]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+  tb.insertNodes(afterSelection.end.path, nodes);
+  editorState.apply(tb);
+}
+
+void handlePaste(EditorState editorState) async {
+  final data = await AppFlowyClipboard.getData();
+
+  if (editorState.selection?.isCollapsed ?? false) {
+    _pastRichClipboard(editorState, data);
+    return;
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:229]:
```suggestion
    return _pastRichClipboard(editorState, data);
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 16]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+    Node(
+      type: node.type,
+      attributes: node.attributes
+        ..remove(ParagraphBlockKeys.delta)
+        ..addAll({ParagraphBlockKeys.delta: nextDelta.toJson()}),
+    )
+  );
+}
+
+void _pastRichClipboard(EditorState editorState, AppFlowyClipboardData data) {
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:271]:
```suggestion
void _pasteRichClipboard(EditorState editorState, AppFlowyClipboardData data) {
```

File: lib/src/service/internal_key_event_handlers/copy_paste_handler.dart [hunk 17]
```diff
@@ -73,3 +82,217 @@ void handlePastePlainText(EditorState editorState, String plainText) {
...
+void deleteSelectedContent(EditorState editorState) async {
+  final selection = editorState.selection?.normalized;
+  if (selection == null || selection.isCollapsed) {
+    return;
+  }
+  final tb = editorState.transaction;
+  await editorState.deleteSelection(selection);
+  tb.afterSelection = Selection.collapsed(selection.start);
+  editorState.apply(tb);
+  return;
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/copy_paste_handler.dart:297]: Redundant return

**Review by Xazin — APPROVED** [review]
Some formatting could be nicer, and some generic syntax.

**Review by MayurSMahajan — APPROVED** [review]

**Pull request merged**